src/target/s32k3xx.c

```diff
@@ -0,0 +1,271 @@
+		spelock_val &= ~(1U << sector);
+		target_mem_write32(flash->t, spelock_reg, spelock_val);
+	} else {
+		/* Use super sector unlock */
+		uint8_t supersector = (addr - flash->start) / SUPER_SECTOR_SIZE;
+		uint32_t sspelock_reg = C40ASF_SSPELOCK_REG(s32flash->block);
+
+		uint32_t sspelock_val = target_mem_read32(flash->t, sspelock_reg);
+		sspelock_val &= ~(1U << supersector);
+		target_mem_write32(flash->t, sspelock_reg, sspelock_val);
+	}
+	return true;
+}
+
+static bool s32k3xx_flash_trigger_mcr(target_flash_s *const flash, uint32_t mcr_bits)
+{
+	uint32_t mcr = target_mem_read32(flash->t, C40ASF_MCR);
+	mcr |= mcr_bits;
+	target_mem_write32(flash->t, C40ASF_MCR, mcr);
+
+	/* Set EVH to trigger operation */
+	mcr |= C40ASF_MCR_EHV;
+	target_mem_write32(flash->t, C40ASF_MCR, mcr);
+
+	/* Wait for DONE to be set.
+	 * According to section 9.1 of S32KXX DS, lifetime max times for:
+	 * Quad-page program: 450 uS 
+	 * 8 KB sector erase: 30 ms (typ 8.5),
+	 * First wait 1 ms, then wait 10 ms at a time until we timeout
+	 */
+	platform_timeout_s wait_timeout;
+	platform_timeout_set(&wait_timeout, 60);
+	platform_delay(1);
+	while (
+		!(target_mem_read32(flash->t, C40ASF_MCRS) & C40ASF_MCRS_DONE) && !platform_timeout_is_expired(&wait_timeout))
+		platform_delay(10);
+
+	if (!(target_mem_read32(flash->t, C40ASF_MCRS) & C40ASF_MCRS_DONE)) {
+		DEBUG_ERROR("MCRS[DONE] not set after operation\n");
+		return false;
+	}
+
+	/* Clear the EVH bit first */
+	mcr = target_mem_read32(flash->t, C40ASF_MCR);
+	mcr &= ~C40ASF_MCR_EHV;
+	target_mem_write32(flash->t, C40ASF_MCR, mcr);
+
+	uint32_t mcrs = target_mem_read32(flash->t, C40ASF_MCRS);
+
+	/* Then clear the operation bits */
+	mcr &= ~mcr_bits;
+	target_mem_write32(flash->t, C40ASF_MCR, mcr);
+
+	if ((mcrs & C40ASF_MCRS_PEG) == 0U) {
+		DEBUG_ERROR("MCRS[PEG] not set after operation\n");
+		return false;
+	}
+
+	if ((mcrs & 0xffff0000U) > 0U) {
+		DEBUG_ERROR("Operation failed, MCRS: %x\n", mcrs);
+		return false;
+	}
+	return true;
+}
+
+static void s32k3xx_flash_prepare(target_flash_s *const flash)
+{
+	uint32_t mcrs = target_mem_read32(flash->t, C40ASF_MCRS);
+	mcrs |= C40ASF_MCRS_PEP | C40ASF_MCRS_PES;
+	target_mem_write32(flash->t, C40ASF_MCRS, mcrs);
+}
+
+static bool s32k3xx_flash_erase(target_flash_s *const flash, target_addr_t addr, size_t len)
+{
+	(void)len;
+	s32k3xx_flash_prepare(flash);
+	s32k3xx_unlock_address(flash, addr);
+
+	target_mem_write32(flash->t, PFCPGM_PEADR_L, addr);
+	target_mem_write32(flash->t, C40ASF_DATA0, 0U);
+	if (!s32k3xx_flash_trigger_mcr(flash, C40ASF_MCR_ERS))
+		return false;
+
+	return true;
+}
+
+static bool s32k3xx_flash_write(target_flash_s *flash, target_addr_t dest, const void *src, size_t len)
+{
+	assert(len == flash->writesize);
+
+	const uint32_t *const s_data = src;
+	s32k3xx_flash_prepare(flash);
+	target_mem_write32(flash->t, PFCPGM_PEADR_L, dest);
+	for (uint32_t i = 0; i < len; i += 4) {
```

Comment:
This loop counter should use `size_t` for its type to match `len` so we don't accidentally generate integer promotions that would turn `i` into a signed value